build-logic/asciidoc-extensions/src/main/java/org/spockframework/plugins/asciidoctor/IncludedSourceLinker.java

```diff
@@ -0,0 +1,257 @@
+package org.spockframework.plugins.asciidoctor;
+
+import org.asciidoctor.ast.*;
+import org.asciidoctor.extension.IncludeProcessor;
+import org.asciidoctor.extension.PreprocessorReader;
+import org.asciidoctor.extension.Treeprocessor;
+import org.asciidoctor.log.LogRecord;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static java.lang.String.join;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Collections.emptyMap;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.*;
+import static org.asciidoctor.extension.Contexts.LISTING;
+import static org.asciidoctor.extension.Contexts.LITERAL;
+import static org.asciidoctor.log.Severity.WARN;
+
+public class IncludedSourceLinker {
```

Comment:
This is quite a lot more complex than the other processor, can we reasonably write unit tests for this?

Reply:
Maybe. I mainly hacked this together quickly and wanted at least first feedback if you like it at all. :-D
Also I hope to get rid of half of the code, when Asciidoctor hopefully exposes a way to reuse the tag filterting logic.
The only way to get the filename of the included file is a custom include processor and with a custom include processor you currently have to do all the include handling yourself.

Reply:
Is there an Asciidoctor issue for this?

Reply:
It indeed was hacked together too quickly and had some problems.
I now did a 1:1 port from the original Ruby code in Asciidoctor to behave exactly the same as of now,
and also added `lines` and `indent` support on the include directive.
Now only `leveloffset` which should not make sense, `encoding` which should not be necessary, and `opts` which likely isn't needed are unsupported with the custom include handling.

As most of the code indeed is the 1:1 port from the Asciidoctor Ruby code with added code comments, that hopefully can be thrown out at some point, I'm not sure whether I really want to write tests for it. :-D

> Is there an Asciidoctor issue for this?

Yes: asciidoctor/asciidoctor#571